### ISSUE TITLE
How to import Train Eval splits when they are pre-split and exist in other locations?

### ISSUE THREAD

**Issue by rclough** [tensorflow/tfx#2496]
From v0.21+, the expected format of an [Examples artifact](https://github.com/tensorflow/tfx/blob/b9a399c46a23b5b433886d4c40d602cd088eb2d8/tfx/types/standard_artifacts.py#L46-L52) is that the Artifact has a base path, which is the URI native to all Artifacts, and then a `split_names` property, which are effectively subfolders of the base URI that contain the data for each split.

This is a bit different from versions <=0.15, where there was a `splits` property, which could contain arbitrary URIs, such that  
for example `train` and `eval` datasets could live in separate locations. The change is in [this commit](https://github.com/tensorflow/tfx/commit/8d3ba19739d553b74e652bffb2ab4a5e5c240247).

This has change proven a little bit cumbersome for some of our teams data engineering/ML setups, where there are scenarios where train and eval data is generated separately, and into separate GCS locations, sometimes due to constraints of our own internal data engineering systems. In addition, there is another common use case for which it is a bit awkward, which is having a standardized evaluation/holdout data set that is consistent across multiple pipeline runs. This is common for us to have an apples-to-apples comparison of changes to models/feature transformation, and evaluating new training data, etc.

In the current state, we have to have a custom component which copies the eval data and the training data into the correct folder layout, and then use an Importer node to load the Examples. While this certainly works and is relatively encapsulated once implemented, it does seem a bit wasteful, especially in the case where we are consistently using the same evaluation dataset for many pipeline runs.

We wanted to open this ticket for guidance on how best to approach this issue, or if maybe there is some kind of functionality that could be added to TFX to support this use case.

**Comment by 1025KB** [comment]
Currently we support import pre-split train and eval by using the [input_config](https://github.com/tensorflow/tfx/blob/master/docs/guide/examplegen.md#custom-inputoutput-split)

Splits will be in a single artifact in ExampleGen's output, and you can access the split by using [this](https://github.com/tensorflow/tfx/blob/2081eee1e94d97de399514e02079a233309d648d/tfx/types/artifact_utils.py#L111)

**Comment by rclough** [comment]
the `input_config` does not support the case described in the original issue: when your pre-split data exists in different GCS paths. The `input_config` makes the assumption that the splits are subdirectories of the Artifact `uri`.

**Comment by 1025KB** [comment]
So the splits don't have a common base dir? (it might work if the base dir is just gs://)

About why we have them in as a single artifact, from Charles: This was done so that the Examples data could be treated as an atomic unit for the purposes of triggering asynchronous executions (e.g. ExampleGen running asynchronously could produce a new artifact for each span of data) and so that they could be coupled for garbage collection.

**Comment by rclough** [comment]
Sorry I dropped the ball on this ticket. I've come back to this issue again after speaking with some of our users recently... It seems having a common base path is a workable approach (even if its just at the bucket or even `gs://`), but there is still at least one concern with the `input_config` approach, which is that it can be  really annoying for large datasets, since using an input_config will force a dataflow shuffle on the data in order to copy it to the new location, which is wasteful and time consuming (admittedly it could be much worse, but imagine you are using TFX to launch multiple variations of a pipeline to compare against a single evaluation dataset)

Before, we could take data that we know was appropriately shuffled and essentially do a no-op to create the Examples Artifact, but now with the way splits are handled, we at least need to force shuffle the data and copy it to a new location.

Perhaps a better approach in this scenario is to just use the `Importer` for two single-split datasets, one for train, and one for eval, and use those accordingly to the correct components? Now that split configuration is more configurable at the component level. Although this still has downsides to components that might make use of both, ie `Trainer`

**Comment by rcrowe-google** [comment]
Is it necessary to preserve the order of the pre-shuffled data?

**Comment by singhniraj08** [comment]
@rclough,

For your shuffling concerns, The shuffling is done within the split, it won't shuffle across splits. If you provide pre-generated train and eval split, the output train and eval is mapped to the input train and eval. 
but for both, we will reshuffle the data within split before materialization (this reshuffle doesn't cause any issue as PCollection doesn't guarantee orders anyway).
Thank you!

**Comment by rcrowe-google** [comment]
This might help - https://github.com/tensorflow/tfx-addons/pull/226#pullrequestreview-1434895951

**Comment by singhniraj08** [comment]
Closing this due to inactivity. Please take a look into the answers provided above, feel free to reopen and post your comments(if you still have queries on this). Thank you!
